packages/react/src/useActionForm.ts

```diff
@@ -106,13 +106,41 @@ type UseActionFormSubmit<F extends ActionFunction<any, any, any, any, any> | Glo
   event?: React.BaseSyntheticEvent<object, any, any> | undefined
 ) => Promise<UseActionFormHookState<F>>;
 
+type ExcludeUndefined<T> = T extends undefined ? never : T;
+
+type ServerSideError<F extends ActionFunction<any, any, any, any, any> | GlobalActionFunction<any>> = F extends ActionFunction<
+  any,
+  any,
+  any,
+  any,
+  any
+>
+  ? {
+      [key in F["modelApiIdentifier"]]?: {
+        [key in
+          | keyof F["selectionType"]
+          | keyof ExcludeUndefined<F["variablesType"]>
+          | ExcludeUndefined<F["modelApiIdentifier"] extends "user" ? "password" : never>]?: { message: string };
+      };
+    }
+  : { [key in keyof ExcludeUndefined<F["variablesType"]>]: { message: string } };
+
+type UseActionFormState<
+  F extends ActionFunction<any, any, any, any, any> | GlobalActionFunction<any>,
+  FormVariables extends FieldValues,
+  FormContext
+> = Omit<UseFormReturn<FormVariables, FormContext>["formState"], "errors"> & {
+  errors: UseFormReturn<FormVariables, FormContext>["formState"]["errors"] & ServerSideError<F>;
+};
+
 export type UseActionFormResult<
   GivenOptions extends OptionsType,
   SchemaT,
   ActionFunc extends ActionFunction<GivenOptions, any, any, SchemaT, any> | GlobalActionFunction<any>,
   FormVariables extends FieldValues,
   FormContext = any
-> = Omit<UseFormReturn<FormVariables, FormContext>, "handleSubmit"> & {
+> = Omit<Omit<UseFormReturn<FormVariables, FormContext>, "handleSubmit">, "formState"> & {
```

Comment:
Can you avoid the double omit with `Omit<xyz, "a" | "b">`

Reply:
:pray: